packages/runtime/container-runtime/src/summary/summarizerNode/summarizerNode.ts

```diff
@@ -193,24 +194,114 @@ export class SummarizerNode implements IRootSummarizerNode {
 	}
 
 	/**
-	 * Complete the WIP summary for the given proposalHandle
+	 * Validates that the in-progress summary is correct, i.e., summarize should have run for all non-skipped
+	 * nodes. This will only be called for the root summarizer node and is called by it recursively on all child nodes.
+	 *
+	 * @returns ValidateSummaryResult which contains a boolean success indicating whether the validation was successful.
+	 * In case of failure, additional information is returned indicating type of failure and where it was.
+	 */
+	public validateSummary(): ValidateSummaryResult {
+		return this.validateSummaryCore(false /* parentSkipRecursion */);
+	}
+
+	/**
+	 * Validates that the in-progress summary is correct for all nodes, i.e., summarize should have run for all
+	 * non-skipped nodes.
+	 * @param parentSkipRecursion - true if the parent of this node skipped recursing the child nodes when summarizing.
+	 * In that case, the children will not have work-in-progress state.
+	 *
+	 * @returns ValidateSummaryResult which contains a boolean success indicating whether the validation was successful.
+	 * In case of failure, additional information is returned indicating type of failure and where it was.
+	 */
+	protected validateSummaryCore(parentSkipRecursion: boolean): ValidateSummaryResult {
+		if (!this.didSummarize(parentSkipRecursion)) {
+			return {
+				success: false,
+				reason: "NodeDidNotSummarize",
+				id: {
+					tag: TelemetryDataTag.CodeArtifact,
+					value: this.telemetryNodeId,
+				},
+				// These errors are usually transient and should go away when summarize is retried.
+				retryAfterSeconds: 1,
+			};
+		}
+		if (parentSkipRecursion) {
+			return { success: true };
+		}
+
+		for (const child of this.children.values()) {
+			const result = child.validateSummaryCore(this.wipSkipRecursion || parentSkipRecursion);
+			// If any child fails, return the failure.
+			if (!result.success) {
+				return result;
+			}
+		}
+		return { success: true };
+	}
+
+	private didSummarize(parentSkipRecursion: boolean): boolean {
```

Comment:
Can we call this `didNotSummarize`? IMO it's easier to miss the `!` than the word `Not`. Up to you. I'm going through a lot of `true false true false` mental gymnastics right now.

Reply:
+1.  And given the backwards nature of the handling of `parentSkipRecursion` flag and likewise `gcDisabled` in the other class -- How about `summarizeWasMissed` and `gcWasMissed`?

Reply:
IMO, ! is a very basic operator and it should not be missed because in that case a whole lot of things won't make sense :-).
I agree with the true false transitions and am open to changing it.

Reply:
I like `summarizeWasMissed` and `gcWasMissed` because it conveys these were skipped as well.

Reply:
changed it.